rewrite-java/src/main/java/org/openrewrite/java/search/SemanticallyEqual.java

```diff
@@ -0,0 +1,184 @@
+/*
+ * Copyright 2020 the original author or authors.
+ * <p>
+ * Licensed under the Apache License, Version 2.0 (the "License");
+ * you may not use this file except in compliance with the License.
+ * You may obtain a copy of the License at
+ * <p>
+ * https://www.apache.org/licenses/LICENSE-2.0
+ * <p>
+ * Unless required by applicable law or agreed to in writing, software
+ * distributed under the License is distributed on an "AS IS" BASIS,
+ * WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+ * See the License for the specific language governing permissions and
+ * limitations under the License.
+ */
+package org.openrewrite.java.search;
+
+import org.openrewrite.*;
+import org.openrewrite.java.JavaProcessor;
+import org.openrewrite.java.tree.*;
+
+import java.util.List;
+import java.util.Objects;
+
+/*
+ * Semantically Equal recursively checks the equality of each element of two ASTs to determine if the two trees are semantically equal.
+ * This is necessary because ASTs are so frequently recreated that merely comparing the IDs of two ASTs is ineffective.
+ * SemanticallyEqual has only been implemented for annotations for now.
+ */
+@Incubating(since = "6.0.0")
+public class SemanticallyEqual {
+
+    protected SemanticallyEqualProcessor processor;
+
+    public SemanticallyEqual() {
+        this.processor = new SemanticallyEqualProcessor();
+    }
+
+    public boolean areEqual(J firstElem, J secondElem) {
+        this.processor.visit(firstElem, secondElem); // returns null, but changes value of class variable isEqual
+        return this.processor.isEqual;
+    }
```

Comment:
As written once `SemanticallyEqualProcessor.isEqual` gets set to `false`, nothing can ever set it to `true` again. So a `SemanticallyEqual` object can only ever be used for a single equality check, after that you can't trust `areEqual()` to be correct.

This problem is avoided if you make this a static utility class. Something like this, perhaps:
```java
public class SemanticallyEqual {
    private SemanticallyEqual() {}
    public static boolean areEqual(J first, J second) {
        SemanticallyEqualProcessor sep = new SemanticallyEqualProcessor();
        sep.visit(first, second);
        return sep.isEqual;
    }
```